Makefile

```
@@ -184,17 +187,22 @@ eclint/lint:
 	$(ECLINT_FILES) | grep -zv ".bats" | xargs -0 -I {} eclint check {}
 	@ echo "[$@]: Project PASSED eclint test!"
 
-python/%: FIND_PYTHON := find . $(FIND_EXCLUDES) -name '*.py' -type f
-## Lints Python files
-python/lint: | guard/program/black
+python/%: PYTHON_FILES := git ls-files '*.py'
+## Checks format and lints Python files.  Runs pylint on each individual
+## file and uses a custom format for the lint messages.
+python/lint: | guard/program/pylint guard/program/black guard/program/git
+python/lint:
 	@ echo "[$@]: Linting Python files..."
-	$(FIND_PYTHON) | $(XARGS) black --check $$(dirname {})
+	$(PYTHON_FILES) | xargs black --check
```

this will fix the test failure:

```suggestion
	$(PYTHON_FILES) | $(XARGS) black --check {}
```

---

No ... that causes black to be rerun twice.  I've fixed it, but I'm not sure I've pushed it as I'm working on the unit tests.

---

I've not been seeing that myself, when running the tests locally. Can you share what you're seeing? Note the test works on two _different_ files, so it is expected to see black run twice here:

```
$ make python/lint
[python/lint]: Linting Python files...
git ls-files --cached --others --exclude-standard '*.py' | xargs -I {} black --check {}
would reformat python_lint_failure/nested/test.py
Oh no! 💥 💔 💥
1 file would be reformatted.
would reformat python_lint_failure/test.py
Oh no! 💥 💔 💥
1 file would be reformatted.
make: *** [../../Makefile:196: python/lint] Error 123
```

---

I'll look at this again tomorrow, er ... later today.

---

Ok, I was wrong again.  I was using a repo with two files that had format issues and when I saw two reports, I made the erroneous leap of logic that black was run twice.  We have a choice of two types of output:

```bash
----- test1 -----------------------------------------------------------
would reformat /home/kbalk/plus3it/tardigrade-modules/junk.py
would reformat /home/kbalk/plus3it/tardigrade-modules/junk2.py
would reformat /home/kbalk/plus3it/tardigrade-modules/junk1.py
would reformat /home/kbalk/plus3it/tardigrade-modules/junk3.py
Oh no! 💥 💔 💥
4 files would be reformatted.
----- test2 -----------------------------------------------------------
would reformat junk.py
Oh no! 💥 💔 💥
1 file would be reformatted.
would reformat junk1.py
Oh no! 💥 💔 💥
1 file would be reformatted.
would reformat junk2.py
Oh no! 💥 💔 💥
1 file would be reformatted.
would reformat junk3.py
Oh no! 💥 💔 💥
1 file would be reformatted.
```

I prefer the first output and I'd really prefer not to see the emojis, but that's not an option.  Which output would you prefer that we use?

---

I prefer the first also. And it turns out black does have an --exclude option, not sure how I missed that before...

https://black.readthedocs.io/en/stable/installation_and_usage.html#command-line-options

So, I'd say set PYTHON_BLACK_EXCLUDE as an overidable make variable (?=) and set the default value the same as the documentation...

Does pylint support an exclude cli option also?

---

I had typed in my comment above and then typed in a comment in Trello, but forgot to hit the "Comment" button, so my Trello comment appeared first.  Sorry about that ... it would have made my Trello comment clearer.

Ok, so yes, pylint does support excludes:

```
    --ignore=<file>[,<file>...]
                        Add files or directories to the blacklist. They should
                        be base names, not paths. [current: CVS]
    --ignore-patterns=<pattern>[,<pattern>...]
                        Add files or directories matching the regex patterns
                        to the blacklist. The regex matches against base
                        names, not paths. [current: none]
```

Would PYTHON_EXCLUDES make more sense or do you want separate black and pylint environment variables for exclusions?

I have been struggling with npm and the installation of editorconfig-checker.  On Centos 7, npm produces errors trying to install the package.  I updated npm and it complained about package-lock.json, but supposedly installed editorconfig-checker.   OSx just installed the package without an issue and yet OSx has an old version of npm.

---

Since they both accept a regex argument (--exclude for black, --ignore-patterns for pylint), yes, PYTHON_EXCLUDES could make sense. As long as the same value works the same for both tools. If the value is interpreted differently by each tool, then as you noted, different envs would be necessary.

---

I also struggle with nodejs/npm. I don't really want a system-level install (e.g. the `-g` argument to npm install), but neither do I want a local install into ./node_modules, which seems to be the only other option. I generally want a user-level install into $HOME. But nodejs doesn't seem to do that very well, or at all, that I've found. At the moment, I say pretty please do what I want, and `sudo <npm install thing>`, and save the fight for another day...

---

This is a go repo so the tool can be compiled with go:  https://github.com/editorconfig-checker/editorconfig-checker

There are also a slew of go binaries available for the tool:  https://github.com/editorconfig-checker/editorconfig-checker/releases

I'll play with the go compilation and see how that goes.

I saw that eclint was deprecated, but it's not a fancy tool and unless there are new file types out there, it does the job.  But I also see the value of moving away from a deprecated tool.

---

Also, I would really prefer that the use of a replacement tool for eclint was a separate ticket.  This task keeps expanding and I'd really like to get the lint stuff out there now.  I'm going to focus on the exclusion flags for Python for now.

---

Completely acceptable. Was only trying to make it easier by getting the broken eclint tests fixed up, so they don't interfere with untracked files and ongoing development happening in this pr.

---

Fixing the tests are a separate task and maybe converting to a different eclint tool could be part of that task?

---

You had asked about combining the tasks, and I thought that was the plan. But it doesn't matter to me in the end. I am tracking both issues. Whatever makes it easiest for you.

---

How about this - I add the python exclude function as the last subtask for this ticket and I'm done with this task.  The eclint-related task will include a eclint replacement and "repairing" the makefile tests for eclint & python linting.  Would that work?

It looks like black changed their handling for excludes late last year - the exclude pattern was for recursive searching and not for files specified on the command line.  I'm still researching, but it appears I need a version of black that has that capability.  I need to see if you guys already have a way of specifying versions equal to a greater than a specific version when you install.

---

Ah, here it is:

```
  --force-exclude TEXT            Like --exclude, but files and directories
                                  matching this regex will be excluded even
                                  when they are passed explicitly as arguments
```

---

The 'force-exclude' option was part of release 20.8b0 of black, made available in August 26, 2020.